File: 0070_Climbing Stairs/ClimbingStairsRecursion.cs
public class Solution {
    Dictionary<int,int> cache = new Dictionary<int,int>();
    public int ClimbStairs(int n) {
        if(n <= 1) return 1;
        if(cache.ContainsKey(n)) return cache[n];
        var r = ClimbStairs(n-1) + ClimbStairs(n-2);
        cache.Add(n,r);
        return r;
    }
}

File: 0070_Climbing Stairs/ClimbingStairsRecursion.java
class Solution {
    public int climbStairs(int n) {
        return climbStairs(n, new HashMap<Integer,Integer>());
    }
    
    private int climbStairs(int n, Map<Integer,Integer> memo) {
        if(n <= 2) return n;
        if(memo.containsKey(n)) return memo.get(n);
        int v = climbStairs(n-1, memo) + climbStairs(n-2, memo);
        memo.put(n,v);
        return v;
    }
}